Frameworks/Foundation/NSInvocation.mm

```diff
@@ -268,3 +277,44 @@ - (void)invoke {
 }
 
 @end
+
+extern "C" void _NSInvocation_ForwardFrame(
+    void* stret, id self, SEL sel, void* frame, /* out */ _NSInvocationForwardReturnInfo* bridgeOut) {
+    NSMethodSignature* signature = [self methodSignatureForSelector:sel];
+    if (!signature) {
+        const char* types = nullptr;
+        types = sel_getType_np(sel);
+        if (!types) {
+            SEL typedSelector = nullptr;
+            sel_copyTypedSelectors_np(sel_getName(sel), &typedSelector, 1);
+            if (typedSelector) {
+                types = sel_getType_np(typedSelector);
+            }
+        }
+
+        if (types) {
+            signature = [NSMethodSignature signatureWithObjCTypes:types];
+        }
+    }
+
+    if (!signature) {
+        Class cls = object_getClass(self);
+        TraceError(TAG,
+                   L"%c[%ls %ls]: unable to find method signature for forwarding.",
+                   class_isMetaClass(cls) ? '+' : '-',
+                   class_getName(cls),
+                   sel_getName(sel));
+        [self doesNotRecognizeSelector:sel]; // This should not return...
+        *bridgeOut = { nullptr, 0 }; // ...but if it does, populate the output anyway.
+        return;
+    }
+
+    // frame is platform-specific; the _NSInvocationCallFrame knows what to do with it.
+    NSInvocation* invocation = [[[NSInvocation alloc] _initWithMethodSignature:signature copyInFrame:frame] autorelease];
+    auto& callFrame = *invocation->_callFrame.get();
+    [self forwardInvocation:invocation];
+    if (stret) {
```

Comment:
Sorry, I don't understand the question.

`stret` gets memcpy'd in through the invocation codepath; other return values are register-loaded by the platform-specific thunks.

---
In reply to: [77276695](https://github.com/Microsoft/WinObjC/pull/951#discussion_r77276695) [](ancestors = 77276695)